OCPBUGS-30343: Scope each MCN object to only be accessible from its associated MCD

djoshy opened the pull request:
This PR adds:
- a ValidatingAdmissionPolicy that limits the service account of a node's MCD from editing the MCN associated with another node
- tests in e2e/techpreview to verify this behavior

How to test manually:
- Launch a cluster in TechPreview
- From a random node's MCD pod, attempt to patch the MCN of another node. This should fail.
```
$ oc rsh pod/machine-config-daemon-l2wmr ./rootfs/usr/bin/oc patch  machineconfignodes/djoshy11-gwphs-worker-
b-b5r6s.c.openshift-gce-devel.internal --type=merge -p '{"spec":{"configVersion":{"desired":"rendered-worker-test"}}}'

The machineconfignodes "djoshy11-gwphs-worker-b-b5r6s.c.openshift-gce-devel.internal" is invalid: : ValidatingAdmissionPolicy 'mcn-guards' with binding 'mcn-guards-binding' denied request: MCN update request from node djoshy11-gwphs-master-0.c.openshift-gce-devel.internal rejected, updates to MCN djoshy11-gwphs-worker-b-b5r6s.c.openshift-gce-devel.internal can only be done by the machine-config-daemon running on the MCN's owner node
```

- Attempt to directly patching the MCN by impersonating the machine-config-daemon SA. This should fail.
```
$ oc patch  machineconfignodes/djoshy11-s9km2-master-0.c.openshift-gce-devel.internal --type=merge -p '{"spec":{"configVersion":{"desired":"rendered-worker-test"}}}' --as='system:serviceaccount:openshift-machine-config-operator:machine-config-daemon'

The machineconfignodes "djoshy11-s9km2-master-0.c.openshift-gce-devel.internal" is invalid: : ValidatingAdmissionPolicy 'mcn-guards' with binding 'mcn-guards-binding' denied request: this user must have a "authentication.kubernetes.io/node-name" claim
```
 
- Attempt to patch the MCN from the MCD running on its node. This should succeed, as it is on the correct node and is using the MCD SA.
``` 
$ oc rsh pod/machine-config-daemon-l2wmr ./rootfs/usr/bin/oc patch  machineconfignodes/djoshy11-gwphs-master-0.c.openshift-gce-devel.internal --type=merge -p '{"spec":{"configVersion":{"desired":"rendered-worker-test"}}}'
machineconfignode.machineconfiguration.openshift.io/djoshy11-gwphs-master-0.c.openshift-gce-devel.internal patched
```
- And last but not least, attempt to patch directly from oc using your admin SA. This should still work as normal as the VAP only applies to requests using the MCD SA.
```
$ oc patch  machineconfignodes/djoshy11-s9km2-master-0.c.openshift-gce-devel.internal --type=merge -p '{"spec":{"configVersion":{"desired":"rendered-worker-test"}}}'
machineconfignode.machineconfiguration.openshift.io/djoshy11-s9km2-master-0.c.openshift-gce-devel.internal patched
```

djoshy commented:
/test e2e-techpreview

djoshy commented:
/test e2e-gcp-op-techpreview

hexfusion commented on test/e2e-techpreview/mcn_test.go:
```diff
@@ -0,0 +1,36 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestMCNScopeSadPath(t *testing.T) {
```
:sob:

JoelSpeed reviewed (commented):
This will need to consider also someone impersonating the service account, [see here](https://github.com/openshift/origin/pull/28673/files#diff-72bbc8936ed74501a136ed7ee2f3cff8f70ac7209ef8d90f87db010f6bfcebeeR134-R139)https://github.com/openshift/origin/pull/28673/files#diff-72bbc8936ed74501a136ed7ee2f3cff8f70ac7209ef8d90f87db010f6bfcebeeR134-R139) for an example of how to test this and then [here](https://github.com/openshift/origin/pull/28673/files#diff-81485c95b8a0742bba2011fea12dccecf47ef6ff17064d114d841e37d2f1103aR21-R22) as an example of protecting against the lack of a node extra

yuqi-zhang commented on manifests/machineconfigdaemon/mcn-guards-validatingadmissionpolicy.yaml:
```diff
@@ -0,0 +1,22 @@
+    resourceRules:
+    - apiGroups:   ["machineconfiguration.openshift.io"]
+      apiVersions: ["v1alpha1"]
+      operations:  ["CREATE","UPDATE"]
```
Q: the daemon seems to also have patch perms. Would this also be needed here?

djoshy replied:
I believe UPDATE encompasses patching as well. VAP only gives you a [smaller "operation" list](https://github.com/djoshy/machine-config-operator/blob/203afde4a93e2c6c7672d6669dd6fe7644ccd388/vendor/k8s.io/api/admissionregistration/v1/types.go#L525) to match on, not sure why :thinking:

yuqi-zhang commented on pkg/operator/sync.go:
```diff
@@ -1359,6 +1361,12 @@ func (optr *Operator) syncMachineConfigDaemon(config *renderConfig) error {
 		configMaps: []string{
 			mcdKubeRbacProxyConfigMapPath,
 		},
+		validatingAdmissionPolicies: []string{
```
probably a dumb question, what is the mechanism by which this is being gated by featuregate?

As in, this will always sync, but if we dont have techpreview on, it will see the object as `apiVersion: admissionregistration.k8s.io/v1beta1` and refuse to apply it...?

djoshy replied:
It's currently gated in applyManifests [here](https://github.com/openshift/machine-config-operator/blob/aa28d1b034bcfaff79537d2fc979de018a77e4a0/pkg/operator/sync.go#L902). VAP does GA in 1.30, so we'll have to update this to respect the actual features at that point. If we potentially time the kube bump with MCN GAing we could skip that middle step

djoshy commented:
> This will need to consider also someone impersonating the service account, [see here](https://github.com/openshift/origin/pull/28673/files#diff-72bbc8936ed74501a136ed7ee2f3cff8f70ac7209ef8d90f87db010f6bfcebeeR134-R139)https://github.com/openshift/origin/pull/28673/files#diff-72bbc8936ed74501a136ed7ee2f3cff8f70ac7209ef8d90f87db010f6bfcebeeR134-R139) for an example of how to test this and then [here](https://github.com/openshift/origin/pull/28673/files#diff-81485c95b8a0742bba2011fea12dccecf47ef6ff17064d114d841e37d2f1103aR21-R22) as an example of protecting against the lack of a node extra

Will update to reflect the impersonation case w/o the extra as well.

djoshy commented:
I've updated the policy to specifically error for the impersonation condition. I'm still using a `matchCondition` in the VAP to only check for the machine-config-daemon SA requests, as I think it is a bit easier to read the policy without having to use variables. I've also added a new test for the impersonation case.

yuqi-zhang reviewed (approved):
Update lgtm based on my understanding. Will let @JoelSpeed take a look

JoelSpeed reviewed (commented):
Nits, otherwise LGTM

djoshy commented:
Fixed the newlines, I'll hold for pre merge QE review

/hold

djoshy commented:
/jira refresh

ptalgulk01 commented:
Verified using this steps:

1. Launched the cluster in TechPreview using Cluster Bot
`launch 4.16,openshift/machine-config-operator#4346 aws,techpreview
`
2. Check machineconfignodes
`oc get machineconfignodes.machineconfiguration.openshift.io 
NAME                           UPDATED   UPDATEPREPARED   UPDATEEXECUTED   UPDATEPOSTACTIONCOMPLETE   UPDATECOMPLETE   RESUMED
ip-10-0-104-242.ec2.internal   True      False            False            False                      False            False
ip-10-0-20-125.ec2.internal    True      False            False            False                      False            False
ip-10-0-6-189.ec2.internal     True      False            False            False                      False            False
ip-10-0-69-160.ec2.internal    True      False            False            False                      False            False
ip-10-0-69-214.ec2.internal    True      False            False            False                      False            False
ip-10-0-98-58.ec2.internal     True      False            False            False                      False            False
`
`oc get po -n openshift-machine-config-operator  -o wide -l k8s-app=machine-config-daemon
NAME                          READY   STATUS    RESTARTS   AGE   IP             NODE                           NOMINATED NODE   READINESS GATES
machine-config-daemon-6bzmp   2/2     Running   0          75m   10.0.20.125    ip-10-0-20-125.ec2.internal    <none>           <none>
machine-config-daemon-85lqz   2/2     Running   0          80m   10.0.69.160    ip-10-0-69-160.ec2.internal    <none>           <none>
machine-config-daemon-blgph   2/2     Running   0          80m   10.0.69.214    ip-10-0-69-214.ec2.internal    <none>           <none>
machine-config-daemon-f22nf   2/2     Running   0          80m   10.0.6.189     ip-10-0-6-189.ec2.internal     <none>           <none>
machine-config-daemon-hm2bn   2/2     Running   0          75m   10.0.104.242   ip-10-0-104-242.ec2.internal   <none>           <none>
machine-config-daemon-qzpss   2/2     Running   0          75m   10.0.98.58     ip-10-0-98-58.ec2.internal     <none>           <none>
`
3.To check not to patch the content of another MCN by another MCD   
`oc rsh -n openshift-machine-config-operator pod/machine-config-daemon-85lqz ./rootfs/usr/bin/oc patch machineconfignodes/ip-10-0-69-214.ec2.internal --type=merge -p '{"spec":{"configVersion":{"desired":"rendered-worker-test"}}}'
Defaulted container "machine-config-daemon" out of: machine-config-daemon, kube-rbac-proxy
The machineconfignodes "ip-10-0-69-214.ec2.internal" is invalid: : ValidatingAdmissionPolicy 'mcn-guards' with binding 'mcn-guards-binding' denied request: updates to MCN ip-10-0-69-214.ec2.internal can only be done from the MCN's owner node
command terminated with exit code 1
`
4. To check not able to patch MCN by MCD's SA
`oc patch machineconfignodes/ip-10-0-69-214.ec2.internal --type=merge -p '{"spec":{"configVersion":{"desired":"rendered-worker-test"}}}' --as='system:serviceaccount:openshift-machine-config-operator:machine-config-daemon' 
The machineconfignodes "ip-10-0-69-214.ec2.internal" is invalid: : ValidatingAdmissionPolicy 'mcn-guards' with binding 'mcn-guards-binding' denied request: this user must have a "authentication.kubernetes.io/node-name" claim
`
5. To check  able to patch the MCN by same MCD running on node
`oc rsh -n openshift-machine-config-operator pod/machine-config-daemon-85lqz ./rootfs/usr/bin/oc patch machineconfignodes/ip-10-0-69-160.ec2.internal --type=merge -p '{"spec":{"configVersion":{"desired":"rendered-worker-test"}}}'
Defaulted container "machine-config-daemon" out of: machine-config-daemon, kube-rbac-proxy
machineconfignode.machineconfiguration.openshift.io/ip-10-0-69-160.ec2.internal patched 
`
6.To check patch directly from oc using your admin SA.
`oc patch machineconfignodes/ip-10-0-69-214.ec2.internal --type=merge -p '{"spec":{"configVersion":{"desired":"rendered-worker-test"}}}' 
machineconfignode.machineconfiguration.openshift.io/ip-10-0-69-214.ec2.internal patched
`

ptalgulk01 commented:
Pre-merge verified https://issues.redhat.com/browse/OCPBUGS-30343
/qe-approve

sinnykumari reviewed (commented):
Nice work David
/lgtm

djoshy commented:
/unhold

djoshy commented:
/test all

hexfusion commented on manifests/machineconfigdaemon/mcn-guards-validatingadmissionpolicy.yaml:
```diff
@@ -0,0 +1,23 @@
+      operations:  ["CREATE","UPDATE"]
+      resources:   ["machineconfignodes/status","machineconfignodes"]
+  validations:
+    # all requests should have a node-name claim, this prevents impersonation of the machine-config-daemon SA.
```
what about machine-config-controller? @djoshy @sergiordlr
